File: src/GitRemoteS3/Enums.hs
-- SPDX-FileCopyrightText: 2023-present Amazon.com, Inc. or its affiliates
--
-- SPDX-License-Identifier: Apache-2.0

-- | URI scheme enum, mirroring the Python @UriScheme@ Enum class.
module GitRemoteS3.Enums
  ( UriScheme (..)
  , uriSchemeText
  ) where

-- | The two supported S3 URI schemes.
--
-- * 'S3'    corresponds to @s3:\/\/@
-- * 'S3Zip' corresponds to @s3+zip:\/\/@
data UriScheme
  = S3      -- ^ Plain git bundle store
  | S3Zip   -- ^ Git bundle store plus a @repo.zip@ archive (for AWS CodePipeline)
  deriving (Eq)

instance Show UriScheme where
  show S3    = "s3"
  show S3Zip = "s3+zip"

-- | Convert a 'UriScheme' to its textual representation.
uriSchemeText :: UriScheme -> String
uriSchemeText = show

File: rust/src/enums.rs
// SPDX-FileCopyrightText: Amazon.com, Inc. or its affiliates
//
// SPDX-License-Identifier: Apache-2.0

/// URI scheme variants understood by git-remote-s3.
///
/// Mirrors the Python `UriScheme` enum in `git_remote_s3/enums.py`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UriScheme {
    /// Plain `s3://` — bundle only.
    S3,
    /// `s3+zip://` — bundle **plus** a `repo.zip` archive (CodePipeline source).
    S3Zip,
}

impl std::fmt::Display for UriScheme {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            UriScheme::S3 => write!(f, "s3"),
            UriScheme::S3Zip => write!(f, "s3+zip"),
        }
    }
}

impl std::str::FromStr for UriScheme {
    type Err = String;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        match s {
            "s3" => Ok(UriScheme::S3),
            "s3+zip" => Ok(UriScheme::S3Zip),
            other => Err(format!("unknown URI scheme: {}", other)),
        }
    }
}
